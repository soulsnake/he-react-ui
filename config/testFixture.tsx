import * as React from 'react';
import createTestContext from 'react-cosmos-test/enzyme';
import toJSON from 'enzyme-to-json';

interface Fixture {
  component: React.ComponentType<any>;
  name?: string;
}

interface EnzymeElement {
  node: {
    nodeType: string;
    props: {
      children?: null;
    };
  };
  children: null | (EnzymeElement | null)[];
}

function sanitiseSubtree(
  element: EnzymeElement | null,
  compositeSeenPreviously: boolean,
): EnzymeElement | null {
  if (!element) return element;
  const { node, children } = element;
  if (!node) return element;
  const { nodeType, props } = node;
  const isComposite = nodeType !== 'host';
  const hasChildrenProp = Boolean(props.children);
  const shouldIncludeChildren =
    !isComposite || !compositeSeenPreviously || hasChildrenProp;
  const compositeSeen = compositeSeenPreviously || isComposite;
  const cleanedChildren: (EnzymeElement | null)[] | null = shouldIncludeChildren
    ? children &&
      children.map((it: EnzymeElement | null) =>
        sanitiseSubtree(it, compositeSeen),
      )
    : [];

  return {
    ...element,
    children: cleanedChildren,
  };
}

function shallowSerialise(enzymeTree: any): EnzymeElement | null {
  return sanitiseSubtree((toJSON(enzymeTree) as any) as EnzymeElement, false);
}

function testOneFixture(fixture: Fixture, variant: string = 'a') {
  const { mount, unmount, getWrapper } = createTestContext({ fixture });
  const { component } = fixture;
  const { displayName, name } = component as any;

  test(`<${displayName || name} /> rendered correctly with ${fixture.name ||
    variant} fixture`, () => {
    mount();
    try {
      expect(shallowSerialise(getWrapper())).toMatchSnapshot();
    } finally {
      unmount();
    }
  });
}

// TODO: use real type guard functions to discriminate one fixture vs an array of them

export default function testAllFixtures(fixtures: any, variant?: string) {
  if (!('forEach' in fixtures)) {
    testOneFixture(fixtures as Fixture, variant);
  } else if (typeof fixtures.forEach === 'function') {
    (fixtures as Fixture[]).forEach(it => testOneFixture(it));
  }
}
