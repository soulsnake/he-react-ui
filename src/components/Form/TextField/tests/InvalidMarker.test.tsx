import fixture from '../fixtures/InvalidMarker.fixture';
import testFixture from '../../../../../config/testFixture';

testFixture(fixture);