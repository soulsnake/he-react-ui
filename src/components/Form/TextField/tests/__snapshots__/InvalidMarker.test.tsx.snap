// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> rendered correctly with InvalidMarker fixture 1`] = `
<TextField
  disabled={false}
  error="Something isn't right"
  id="demo"
  inline={false}
  isValid={false}
  isVerified={false}
  label="First Name"
  marker="tick-cross"
  markerTooltip={null}
  name="demo"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onMarkerClick={[Function]}
  value="Bruce"
>
  <div
    className="outer invalid hasMarker"
  >
    <div
      className="block"
    >
      <input
        className="input"
        disabled={false}
        id="demo"
        name="demo"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="text"
        value="Bruce"
      />
      <div
        className="helper"
      >
        <Cross
          className="marker"
          fill="currentColor"
          height={32}
          onClick={[Function]}
          width={32}
        />
      </div>
      <label
        className="label floating"
        htmlFor="demo"
      >
        First Name
      </label>
    </div>
    <label
      className="description"
      htmlFor="demo"
    >
      Something isn't right
    </label>
  </div>
</TextField>
`;
